### DISCUSSION TITLE
[Question][Hw1] How can I get the argument of function call?

### DISCUSSION THREAD

**Discussion by Calmadeas** [prosyslab-classroom/cs424-program-reasoning#395] [Q&A]
To implement the transfer of case print, I want to know the argument of function call instruction `instr` which calls `print`. But I couldn't find a way to reach it. How can I?

**Comment by Hhro** [comment]
Hi,
You can use `Llvm.operand` to parse the operand from instruction.
(To find more APIs, follow this link: https://llvm.moe/ocaml/Llvm.html)
For example, `Llvm.operand instr 0` will return the first operand for the `instr`.
Thanks.

**Comment by Calmadeas** [comment]
> Hi, You can use `Llvm.operand` to parse the operand from instruction. (To find more APIs, follow this link: https://llvm.moe/ocaml/Llvm.html) For example, `Llvm.operand instr 0` will return the first operand for the `instr`. Thanks.

Thank you for your answer.
Let me clarify my question. I thought I have to get the 'argument of function call', which is the integer to `print`, in this case.
`Llvm.operand instr 0`, gave me something I can't understand.
`Llvm.operand instr 1`, gave me the function print, which is the 'operand' of the instruction `call`.
`Llvm.operand instr 2`, gave me segmentation fault.
Am I misunderstanding something?

**Comment by Hhro** [comment]
Ah, I understand your question.
You can parse the function's arguments the same as the instruction's operands.

Let's consider the function of `n` arguments is called, and we want to parse the arguments from it.
In this case, LLVM instruction would seem like `%call = call @fnc(%a_1, %a_2, %a_3, ..., %a_n)`.
Internally, this instruction consists of the `call` opcode and the `n+1` operands, which sequentially represent `n` arguments and the function pointer that references `fnc`.
Therefore, to parse the **k-th** argument from this instruction, one can use `Llvm.operand instr (k-1)`.

To give a more specific example, when the instruction is `call void @print(i32 %add4)` and you want to parse the `%add4`, `Llvm.operand instr 0` will give you that.

Hope this helps.
